### PR TITLE
sync all resources to cluster fields

### LINKED ISSUE #2719: Removing cluster do not removes objects from k8s
Please, answer some short questions which should help us to understand your problem / question better?

- **Which image of the operator are you using?** ghcr.io/zalando/postgres-operator:v1.12.2
- **Where do you run it - cloud or metal? Kubernetes or OpenShift?** Bare Metal K8s
- **Are you running Postgres Operator in production?** no
- **Type of issue?** Bug report

Something wrong with getting proper namespace to process deletion of resources:
- `example-cluster-config` Service & Endpoint
- Postgres Pods (leaved without StatefulSet)

Log of creation and deletion of the cluster:
```
time="2024-08-06T13:15:00Z" level=debug msg="added new cluster: \"example-cluster/example-cluster\"" pkg=controller
time="2024-08-06T13:15:00Z" level=info msg="SYNC event has been queued" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:15:00Z" level=info msg="creating pod service account \"postgres-pod\" in the \"example-cluster\" namespace" pkg=controller
time="2024-08-06T13:15:00Z" level=info msg="ADD event has been queued" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:15:00Z" level=info msg="successfully deployed the pod service account \"postgres-pod\" to the \"example-cluster\" namespace" pkg=controller
time="2024-08-06T13:15:00Z" level=info msg="Creating the role binding \"postgres-pod\" in the \"example-cluster\" namespace" pkg=controller
time="2024-08-06T13:15:00Z" level=info msg="successfully deployed the role binding for the pod service account \"postgres-pod\" to the \"example-cluster\" namespace" pkg=controller
time="2024-08-06T13:15:00Z" level=info msg="syncing of the cluster started" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:15:00Z" level=debug msg="team API is disabled" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=debug msg="team API is disabled" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=info msg="syncing secrets" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=debug msg="created new secret example-cluster/postgres.example-cluster.credentials.postgresql.acid.zalan.do, namespace: example-cluster, uid: 2fd7d74a-fc06-464b-b9e7-77410192850a" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=debug msg="created new secret example-cluster/standby.example-cluster.credentials.postgresql.acid.zalan.do, namespace: example-cluster, uid: 50585020-e07e-484c-8fbb-a8e58a520310" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=debug msg="syncing master service" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=info msg="could not find the cluster's master endpoint" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:00Z" level=warning msg="master is not running, generated master endpoint does not contain any addresses" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:01Z" level=info msg="created missing master endpoint \"example-cluster/example-cluster\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:01Z" level=info msg="could not find the cluster's master service" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:02Z" level=info msg="created missing master service \"example-cluster/example-cluster\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:02Z" level=debug msg="syncing replica service" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:02Z" level=info msg="could not find the cluster's replica endpoint" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:03Z" level=info msg="created missing replica endpoint \"example-cluster/example-cluster-repl\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:03Z" level=info msg="could not find the cluster's replica service" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=info msg="created missing replica service \"example-cluster/example-cluster-repl\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=debug msg="syncing volumes using \"pvc\" storage resize mode" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=info msg="volume claims do not require changes" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=debug msg="syncing statefulsets" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=info msg="cluster's statefulset does not exist" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=info msg="standby cluster streaming from remote primary" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=warning msg="operator only supports standby clusters with 1 pod" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:04Z" level=debug msg="created new statefulset \"example-cluster/example-cluster\", uid: \"3ae8c3ed-cac6-4588-8147-3265f0faa617\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:15:28Z" level=debug msg="Waiting for 3 pods to become ready" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:17:38Z" level=info msg="DELETE event has been queued" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:17:38Z" level=debug msg="event ADD has been discarded for the cluster" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:20:38Z" level=info msg="ADD event has been queued" cluster-name=example-namespace/example-cluster pkg=controller worker=2
time="2024-08-06T13:20:38Z" level=info msg="creating a new Postgres cluster" cluster-name=example-namespace/example-cluster pkg=controller worker=2
time="2024-08-06T13:20:38Z" level=info msg="endpoint \"example-namespace/example-cluster\" has been successfully created" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="master service \"example-namespace/example-cluster\" has been successfully created" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="replica service \"example-namespace/example-cluster-repl\" has been successfully created" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=debug msg="team API is disabled" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=debug msg="team API is disabled" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="users have been initialized" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="syncing secrets" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=debug msg="created new secret example-namespace/postgres.example-cluster.credentials.postgresql.acid.zalan.do, namespace: example-namespace, uid: ad90c461-e228-45e7-9255-73fda3fcbeb8" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="secrets have been successfully created" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="pod disruption budget \"example-namespace/postgres-example-cluster-pdb\" has been successfully created" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="standby cluster streaming from remote primary" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=warning msg="operator only supports standby clusters with 1 pod" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=debug msg="created new statefulset \"example-namespace/example-cluster\", uid: \"93160df7-137a-4198-a739-1fce05cb59c9\"" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="statefulset \"example-namespace/example-cluster\" has been successfully created" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:38Z" level=info msg="waiting for the cluster being ready" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:20:56Z" level=debug msg="Waiting for 3 pods to become ready" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:22:21Z" level=info msg="DELETE event has been queued" cluster-name=example-namespace/example-cluster pkg=controller worker=2
time="2024-08-06T13:25:25Z" level=warning msg="error while syncing cluster state: could not sync statefulsets: cluster is not ready: pod labels error: still failing after 200 retries" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=warning msg="could not set cluster status: could not update status: postgresqls.acid.zalan.do \"example-cluster\" not found" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=error msg="could not sync cluster: could not sync statefulsets: cluster is not ready: pod labels error: still failing after 200 retries" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:25:25Z" level=info msg="deletion of the cluster started" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:25:25Z" level=info msg="removing the logical backup job" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=warning msg="could not remove the logical backup k8s cron job; an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=debug msg="deleting statefulset" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=debug msg="statefulset \"example-cluster/example-cluster\" has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=info msg="statefulset \"example-cluster/example-cluster\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:25Z" level=debug msg="deleting pods" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="no pods to delete" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting PVCs" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="no PVCs to delete" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting secret \"example-cluster/postgres.example-cluster.credentials.postgresql.acid.zalan.do\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="secret \"example-cluster/postgres.example-cluster.credentials.postgresql.acid.zalan.do\" has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="secret \"example-cluster/postgres.example-cluster.credentials.postgresql.acid.zalan.do\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting secret \"example-cluster/standby.example-cluster.credentials.postgresql.acid.zalan.do\"" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="secret \"example-cluster/standby.example-cluster.credentials.postgresql.acid.zalan.do\" has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="secret \"example-cluster/standby.example-cluster.credentials.postgresql.acid.zalan.do\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting pod disruption budget" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="there is no pod disruption budget in the cluster" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting endpoint" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="master endpoint has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="master endpoint \"example-cluster/example-cluster\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting service master" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="master service has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="master service \"example-cluster/example-cluster\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting endpoint" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="replica endpoint has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="replica endpoint \"example-cluster/example-cluster-repl\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="deleting service replica" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="replica service has already been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="replica service \"example-cluster/example-cluster-repl\" has been deleted" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="removing leftover Patroni objects (endpoints / services and configmaps)" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="deleting connection pooler spilo-role=master" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="no connection pooler to delete" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="deleting connection pooler spilo-role=replica" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=debug msg="no connection pooler to delete" cluster-name=example-cluster/example-cluster pkg=cluster
time="2024-08-06T13:25:26Z" level=info msg="cluster has been deleted" cluster-name=example-cluster/example-cluster pkg=controller worker=0
time="2024-08-06T13:30:53Z" level=error msg="failed to create cluster: pod labels error: still failing after 200 retries" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=warning msg="cluster created failed: pod labels error: still failing after 200 retries" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=warning msg="could not set cluster status: could not update status: postgresqls.acid.zalan.do \"example-cluster\" not found" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=error msg="could not create cluster: pod labels error: still failing after 200 retries" cluster-name=example-namespace/example-cluster pkg=controller worker=2
time="2024-08-06T13:30:53Z" level=info msg="deletion of the cluster started" cluster-name=example-namespace/example-cluster pkg=controller worker=2
time="2024-08-06T13:30:53Z" level=info msg="removing the logical backup job" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=warning msg="could not remove the logical backup k8s cron job; an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="deleting statefulset" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=info msg="statefulset \"example-namespace/example-cluster\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="deleting pods" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="no pods to delete" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="deleting PVCs" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="no PVCs to delete" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="deleting secret \"example-namespace/postgres.example-cluster.credentials.postgresql.acid.zalan.do\"" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=info msg="secret \"example-namespace/postgres.example-cluster.credentials.postgresql.acid.zalan.do\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="deleting secret \"example-namespace/standby.example-cluster.credentials.postgresql.acid.zalan.do\"" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="secret \"example-namespace/standby.example-cluster.credentials.postgresql.acid.zalan.do\" has already been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=info msg="secret \"example-namespace/standby.example-cluster.credentials.postgresql.acid.zalan.do\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=debug msg="deleting pod disruption budget" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:53Z" level=info msg="pod disruption budget \"example-namespace/postgres-example-cluster-pdb\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="deleting endpoint" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=info msg="master endpoint \"example-namespace/example-cluster\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="deleting service master" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=info msg="master service \"example-namespace/example-cluster\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="deleting endpoint" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="there is no replica endpoint in the cluster" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="deleting service replica" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=info msg="replica service \"example-namespace/example-cluster-repl\" has been deleted" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="removing leftover Patroni objects (endpoints / services and configmaps)" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=warning msg="could not fetch endpoint \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=info msg="deleting connection pooler spilo-role=master" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="no connection pooler to delete" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=info msg="deleting connection pooler spilo-role=replica" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=debug msg="no connection pooler to delete" cluster-name=example-namespace/example-cluster pkg=cluster worker=2
time="2024-08-06T13:30:56Z" level=info msg="cluster has been deleted" cluster-name=example-namespace/example-cluster pkg=controller worker=2
```

### PR DESCRIPTION
**Pull request by FxKu**
While developing owner references support I realized once again that we store many child resources as cluster objects so we can access them even after the manifest is gone. But not for all resources, which might cause issues when deleting them because c.Namespace field might be empty:

```
time="2024-08-01T12:32:33Z" level=debug msg="removing leftover Patroni objects (endpoints / services and configmaps)" cluster-name=default/acid-clone-cluster pkg=cluster worker=2
time="2024-08-01T12:32:33Z" level=warning msg="could not fetch service \"/\": an empty namespace may not be set when a resource name is provided" cluster-name=default/acid-clone-cluster pkg=cluster worker=2
```

Fixing it for missing resources such as logical backup cron job and streams is easy. But it gets a bit tricky with objects created by Patroni, namely config service/endpoint and optionally configmaps. This PR introduces extra sync treatment for these, but only mapping to cluster fields. No diffs, no updates. This should allow us to loop over them on deletion.

Some more minor changes:
- we don't need to init the clusters kubeResources with nil
- bugfix: patched secrets and endpoints are now synced in cluster fields
- Patroni endpoints are treated separately to master/replica endpoints (only config service reused existing code, hence the extra PostgresRole)
- have only one function to determine the service/endpoint name
- small refactoring of code when services/endpoints are already deleted
- deleteSecret does not require passing the secret, we fetch it from cluster.Secrets map via uid
- if second parameter from a range loop is not needed we don't need to write an underscore (idx, _ := range -> idx := range)

The streams feature saw some bigger refactoring:
- createOrUpdateStreams was turned into syncStream for a single stream CRD which is more in line with the rest of the code
- loop over appIds is now done in syncStreams as well as the deletion of streams that got removed from the manifest (and, thus, missing in appIds list)
- renamed gatherApplicationIds function to more suitable name getDistinctApplicationIds
- deleteStream does not require passing stream, we fetch it from cluster.Streams map via appId
- the recently added check for existing slots (#2704) to allow FES creation is now comparing desired databaseSlots with actual slotsToSync. The previous comparison was not complete: appId loop with index lookup on c.Spec.Streams, which can be bigger, also with a different order.

fixes #2719

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -1199,10 +1209,10 @@ func (c *Cluster) Delete() error {
 		}
 	}
 
-	if err := c.deletePatroniClusterObjects(); err != nil {
+	if err := c.deletePatroniResources(); err != nil {
```
Comment (hughcapet) [inline pkg/cluster/cluster.go:1212]: shouldn't we skip it if `!c.patroniUsesKubernetes()`?

Comment (FxKu) [reply pkg/cluster/cluster.go:1212]: Good question. In the main branch this code is there to not do "blind" K8s API calls against resources that do not exist in infrastructure. In this PR we only delete based on the cluster struct. Therefore, we should only delete what is also allowed to exist.

Comment (hughcapet) [reply pkg/cluster/cluster.go:1212]: just to avoid unnecessary code execution:)

Comment (FxKu) [reply pkg/cluster/cluster.go:1224]: With this delete code there will be no unnecessary API calls at least. One of either ConfigMaps and Endpoints will always get created - no matter what patroniUsesKubernetes() returns (when called inside patroniKubernetesUseConfigMaps). Yes, without right config this leads to errors during create.

File: pkg/cluster/sync.go [hunk 2]
```diff
@@ -173,6 +177,157 @@ func (c *Cluster) syncFinalizer() error {
...
+			c.logger.Infof("new %s endpoints's owner references do not match the current ones", endpointName)
+			ep.ObjectMeta.OwnerReferences = desiredOwnerRefs
+			c.setProcessName("updating %s endpoint", endpointName)
+			ep, err = c.KubeClient.Endpoints(c.Namespace).Update(context.TODO(), ep, metav1.UpdateOptions{})
+			if err != nil {
+				return fmt.Errorf("could not update %s endpoint: %v", endpointName, err)
+			}
+			c.PatroniEndpoints[suffix] = ep
+		}
+		desiredAnnotations := c.annotationsSet(ep.Annotations)
```
Comment (hughcapet) [inline pkg/cluster/sync.go:270]:
```suggestion
		desiredAnnotations := c.annotationsSet(nil)
```

File: pkg/cluster/sync.go [hunk 3]
```diff
@@ -173,6 +177,157 @@ func (c *Cluster) syncFinalizer() error {
...
+			c.logger.Infof("new %s service's owner references do not match the current ones", serviceName)
+			svc.ObjectMeta.OwnerReferences = desiredOwnerRefs
+			c.setProcessName("updating %v service", serviceName)
+			svc, err = c.KubeClient.Services(c.Namespace).Update(context.TODO(), svc, metav1.UpdateOptions{})
+			if err != nil {
+				return fmt.Errorf("could not update %s endpoint: %v", serviceName, err)
+			}
+			c.Services[Patroni] = svc
+		}
+		desiredAnnotations := c.annotationsSet(svc.Annotations)
```
Comment (hughcapet) [inline pkg/cluster/sync.go:311]:
```suggestion
		desiredAnnotations := c.annotationsSet(nil)
```

**Comment by hughcapet** [comment]
not true anymore?
https://github.com/zalando/postgres-operator/blob/a87307e56b264a7e107b1c52c6b4ded5eea8f0eb/e2e/tests/test_e2e.py#L1678

(+ add Patroni objects to `check_cluster_child_resources_owner_references()` in e2e and add config service/endpoints and configmaps to the `TestInheritedAnnotations` unit test)

**Review by hughcapet — DISMISSED** [review]

**Review by hughcapet — COMMENTED** [review]
comment in test_owner_references e2e test is still not adjusted (patroni resources are affected by owner reference)

**Review by hughcapet — APPROVED** [review]

**Review by idanovinda — APPROVED** [review]

**Comment by FxKu** [comment]
:+1:

**Comment by hughcapet** [comment]
👍

**Comment by hughcapet** [comment]
👍

**Pull request merged**